src/core/builders/create-form-validator.spec.ts

```diff
@@ -847,4 +847,122 @@ describe("createFormValidator", () => {
     ]);
   });
+
+  it("dependency change should trigger validation run", async () => {
+    const required = (x: any) => (x ? null : "REQUIRED");
+
+    const { validate } = createFormValidator(Schema, validate => [
+      validate({
+        field: Schema.string,
+        rules: _number => [required],
+        dependencies: [Schema.number],
+      }),
+      validate({
+        field: Schema.number,
+        rules: () => [required],
+      }),
+    ]);
+
+    const getValue = () => "" as any;
+
+    const validation = await validate({ fields: [Schema.number], getValue });
+
+    expect(validation).toEqual([
+      { field: Schema.number, error: "REQUIRED" },
+      { field: Schema.string, error: "REQUIRED" },
+    ]);
+  });
+
+  it("dependency change should trigger validation run for with array.nth", async () => {
+    const { validate } = createFormValidator(Schema, validate => [
+      validate({
+        field: Schema.arrayObjectString.nth,
+        rules: _ => [x => (x.str === "" ? "REQUIRED" : null)],
+        dependencies: [Schema.choice],
+      }),
+      validate(Schema.choice, x => (x === "A" ? "INVALID_VALUE" : null)),
+    ]);
+
+    const getValue = (field: FieldDescriptor<any>): any => {
+      switch (impl(field).__path) {
+        case "arrayObjectString":
+          return [{ str: "sm" }, { str: "" }, { str: "valid string" }];
+        case "arrayObjectString[0]":
+          return { str: "sm" };
+        case "arrayObjectString[1]":
+          return { str: "" };
+        case "arrayObjectString[2]":
+          return { str: "valid string" };
+        case "choice":
+          return "A";
+      }
+    };
+
+    const validation = await validate({
+      fields: [Schema.choice],
+      getValue,
+    });
+
+    expect(validation).toEqual([
+      { field: Schema.choice, error: "INVALID_VALUE" },
+      { field: Schema.arrayObjectString.nth(0), error: null },
+      { field: Schema.arrayObjectString.nth(1), error: "REQUIRED" },
+      { field: Schema.arrayObjectString.nth(2), error: null },
+    ]);
+  });
+
+  it("dependency change should not duplicate validation", async () => {
+    const required = (x: any) => (x ? null : "REQUIRED");
+
+    const { validate } = createFormValidator(Schema, validate => [
+      validate({
+        field: Schema.string,
+        rules: _number => [required],
+        dependencies: [Schema.number],
+      }),
+      validate({
+        field: Schema.number,
+        rules: () => [required],
+      }),
+    ]);
+
+    const getValue = () => "" as any;
+
+    const validation = await validate({
+      fields: [Schema.number, Schema.string],
+      getValue,
+    });
+
+    expect(validation).toEqual([
+      { field: Schema.number, error: "REQUIRED" },
+      { field: Schema.string, error: "REQUIRED" },
+    ]);
+  });
+
+  it("dependency change should trigger validation run even if root field triggers don't match", async () => {
+    const required = (x: any) => (x ? null : "REQUIRED");
```

Comment:
is validation caused by dependency change is still controlled but the triggers of the affected field? (`Schema.string` in this case)
would be nice to cover this in tests as well

Reply:
added `Schema.choice` validation to this test case to test scenario:
1. trigger `Schema.number` validation with `trigger: change`
2. as `Schema.number` is a dependency of `Schema.choice`, `Schema.choice` is then picked up to be validated (even if `Schema.number` validation is skipped due to trigger mismatch)
3. because `Schema.choice` has `triggers: ["blur"]`, its validation is skipped due to trigger mismatch